index.bs

```diff
@@ -1468,19 +1512,186 @@ To <dfn>validate a suffix</dfn> |suffix|, run the following steps:
+   For example `C:\Documents and Settings\username\My Documents`, `/Users/username/Documents`, or `/home/username/Documents`.
+: <dfn enum-value>"downloads"</dfn>
+:: Directory where downloaded files would typically be stored.
+   For example `C:\Documents and Settings\username\Downloads`, `/Users/username/Downloads`, or `/home/username/Downloads`.
+: <dfn enum-value>"music"</dfn>
+:: Directory where audio files would typically be stored.
+   For example `C:\Documents and Settings\username\My Documents\My Music`, `/Users/username/Music`, or `/home/username/Music`.
+: <dfn enum-value>"pictures"</dfn>
+:: Directory where photos and other still images would typically be stored.
+   For example `C:\Documents and Settings\username\My Documents\My Pictures`, `/Users/username/Pictures`, or `/home/username/Pictures`.
+: <dfn enum-value>"videos"</dfn>
+:: Directory where videos/movies would typically be stored.
+   For example `C:\Documents and Settings\username\My Documents\My Videos`, `/Users/username/Movies`, or `/home/username/Videos`.
+
+<div algorithm>
+To <dfn>determine the directory the picker will start in</dfn>, given an optional [=string=] |id|,
+an optional {{StartInDirectory}} |startIn| and an [=environment settings object=] |environment|,
+run the following steps:
+
+1. If |id| given, and is not a [=valid path id=], throw a {{TypeError}}.
+1. If |id|'s [=string/length=] is more than 32, throw a {{TypeError}}.
+
+1. Let |origin| be |environment|'s [=environment settings object/origin=].
+
+1. If |startIn| is a {{FileSystemHandle}}:
+  1. Let |entry| be |startIn|'s [=FileSystemHandle/entry=].
+  1. If |entry| does not represent an [=/entry=] in an [=origin private file system=]:
+    1. If |entry| is a [=file entry=],
+       return the path on the local file system corresponding to the parent directory of |entry|,
+       if such a path can be determined.
+    1. Else, if |entry| is a [=directory entry=],
+       return the path on the local file system corresponding to |entry|,
+       if such a path can be determined.
+
+1. If |id| is non-empty:
+  1. If [=recently picked directory map=][|origin|] [=map/exists=]:
+    1. Let |path map| be [=recently picked directory map=][|origin|].
+    1. If |path map|[|id|] [=map/exists=], then return |path map|[|id|].
+
+1. If |startIn| is a {{WellKnownDirectory}}:
+  1. Return a user agent defined path corresponding to the {{WellKnownDirectory}} value of |startIn|.
+
+1. If |id| is not specified, or is an empty string:
+  1. If [=recently picked directory map=][|origin|] [=map/exists=]:
+    1. Let |path map| be [=recently picked directory map=][|origin|].
+    1. If |path map|[""] [=map/exists=], then return |path map|[""].
+
+1. Return a default path in a user agent specific manner.
+
+</div>
+
+<div algorithm>
+To <dfn>remember a picked directory</dfn>, given an optional [=string=] |id|,
+an [=/entry=] |entry|, and an [=environment settings object=] |environment|,
```

Comment:
Nit: Is there an advantage to passing the environment in, rather than the origin?

(Doesn't really matter, but if this were code we'd want to pass in the more scoped object)

Reply:
Passing in the environment rather than just what we're using as key today seemed more flexible/future proof (probably even in code). I.e. if we'd want to change from origins to storage keys or something like that, that is now completely isolated to these algorithms, rather than spread out across all its callers. So passing in the environment means that callers don't have to worry about what is used as key.